### PR TITLE
Fix use of **/ in .dockerignore

### LINKED ISSUE #29014: Dockerignore multi-dir wildcard issue
## Description

The multi-dir wildcard does not behave like described in the documentation. In case a particular file should be ignored, the pattern matches every file which ends with the same name also.

## Steps to reproduce the issue:
1. Create a file .example and some.example in the root or any subdirectory
2. Add "**/.example" to you .dockerignore
3. Create a Dockerfile which includes a copy command such as `COPY . /`
2. Build the image

## Describe the results you received:
Neither .example nor some.example is added to the image

## Describe the results you expected:
some.example should be present in the image

## Additional information:

In case you work with gradle projects, this behavior is quite important. (e.g. exclude .gradle but not settings.gradle or even build.gradle)

This issue seems to be present since Oct 14, 2015 (eddb14a). As I can see it, this is when the multi-dir wildcard was introduced.

File `pkg/fileutils/fileutils.go`:

Line 172 states: Treat **/ as ** so eat the "/"

Furthermore the documentation for this is quite misleading:

```
Matching is done using Go’s filepath.Match rules. A preprocessing step removes
leading and trailing whitespace and eliminates . and .. elements using Go’s
filepath.Clean. Lines that are blank after preprocessing are ignored.
```

Line 135 on the other hand states that the filepath.Match is not used at all because of "some really weird rules". Line 141 makes use of its Error Message only.

### PR DESCRIPTION
**Pull request by duglin**
img_1116

Closes #29014

Signed-off-by: Doug Davis <dug@us.ibm.com>

### CODE REVIEW

**Comment by tonistiigi** [comment]
LGTM

**Comment by vieux** [comment]
I see the test have changed. is this a breaking change ?

**Comment by duglin** [comment]
Its not so much a breaking change as much as its a fix for a bug.
The case of the `**/` pattern when used against `dir/file` wasn't quite right before.

It popped up because `**/` is supposed to be treated just like `**`. And when `**` appeared at the end of the pattern it was supposed to be converted to a regex `.*`.  However, the check for a `/` after `**` was done after the "end of string/pattern" check, so when we removed the trailing `/` it was too late to notice the EOS - and it got converted into `.*/` or `[^/]*` instead of `.*`.  Hopefully that rambling made sense.

File: pkg/fileutils/fileutils.go [hunk 1]
```diff
@@ -161,17 +161,19 @@ func regexpMatch(pattern, path string) (bool, error) {
...
 					regStr += ".*"
 				} else {
 					// is "**"
-					regStr += "((.*" + escSL + ")|([^" + escSL + "]*))"
-				}
-
-				// Treat **/ as ** so eat the "/"
-				if string(scan.Peek()) == sl {
-					scan.Next()
+					// Note that this allows for any # of /'s (even 0) because
```
Comment (vdemeester) [inline pkg/fileutils/fileutils.go]: Could we have a test to put that in the light in the tests ? :angel:

**Review by vdemeester — APPROVED** [review]
LGTM 🐸
Just a nit

**Comment by justincormack** [comment]
Please can you write a real commit message it makes life incredibly difficult when looking through commits later as then you have to look up #29014 to see what the issue actually is about.

**Comment by duglin** [comment]
@justincormack done

File: pkg/fileutils/fileutils.go (hunk 1 above)
Comment (duglin) [inline pkg/fileutils/fileutils.go]: @vdemeester I think the unit tests cover this:  https://github.com/docker/docker/pull/29043/files#diff-0e16a9c00f7cbaa739ca87d3ec3b1d37R321 
Did I miss a situation in there?

**Comment by LK4D4** [comment]
LGTM

**Pull request merged**